sql/2020/07_SEO/pages_markup_by_device_and_img_loading.sql

```diff
@@ -0,0 +1,107 @@
+#standardSQL
+# pages markup metrics grouped by device and image loading attributes
+
+# helper to create percent fields
+CREATE TEMP FUNCTION AS_PERCENT (freq FLOAT64, total FLOAT64) RETURNS FLOAT64 AS (
+  ROUND(SAFE_DIVIDE(freq, total), 4)
+);
+
+# returns all the data we need from _markup
+CREATE TEMPORARY FUNCTION get_markup_info(markup_string STRING)
+RETURNS STRUCT<  
+  loading ARRAY<STRING>
+> LANGUAGE js AS '''
+var result = {};
+
+//Function to retrieve only keys if value is >0
+function getKey(dict){
+  const arr = [],
+  obj = Object.keys(dict);
+  for (var x in obj){
+    if(dict[obj[x]] > 0){
+      arr.push(obj[x]);
+    }
+  }
+  return arr;
+}
+
+try {
+    var markup;
+
+    if (true) { // LIVE = true
+      markup = JSON.parse(markup_string); // LIVE
+    }
+    else 
+    {
+      // TEST
+      markup = {
+        "images": {
+          "picture": {
+            "total": 0
+            },
+          "source": {
+            "total": 0,
+            "src_total": 0,
+            "srcset_total": 0,
+            "media_total": 0,
+            "type_total": 0
+            },
+          "img": {
+            "total": 37,
+            "src_total": 37,
+            "srcset_total": 7,
+            "alt": {
+              "missing": 1,
+              "blank": 36,
+              "present": 0
+          },
+            "loading": {
+            "auto": Math.floor(Math.random() * 3),
+            "lazy": Math.floor(Math.random() * 3),
+            "eager": Math.floor(Math.random() * 3),
+            "invalid": Math.floor(Math.random() * 3),
+            "missing": Math.floor(Math.random() * 30),
+            "blank": Math.floor(Math.random() * 3)
+            }
+          }
+        }
+      };
+    }
+
+    if (Array.isArray(markup) || typeof markup != 'object') return result;
+
+    if (markup.images) {
+
+      result.loading = getKey(markup.images.img.loading);
+
+    }
+
+} catch (e) {}
+return result;
+''';
+
+SELECT
+  client,
+  loading, 
+total, 
+COUNT(*) AS count,
+AS_PERCENT(COUNT(*), total) AS pct
```

Comment:
Nit but I've always used `COUNT(0)` rather than `COUNT(*)`. Dunno whether still true but was taught `COUNT(*)` required a tabled definition lookup to get the columns, which then weren't even used.

Doubtful it's even noticeable but from a convention it seems others use `COUNT(0)` so best to do a search and replace and replace these IMHO.

Reply:
I'd never heard of COUNT(0) before this. Will search and replace.

Reply:
`COUNT(0)`, `COUNT(1)`, `COUNT('BOB')` all do the same thing. It's just COUNT needs an argument.

`COUNT(column)` will count the number of non-null instances of that column in the rows so is different if some rows have a null value in column. But a static or `*` will return the same info.

And I've always used `COUNT(1)` cause it just reads better to me (count 1 for each row) than `COUNT(0)` but looks like convention here is `COUNT(0)`.

More info here: https://stackoverflow.com/questions/18291036/what-is-the-difference-between-count0-count1-and-count-in-mysql-sql including this bit:

> In days gone by, though, COUNT(1) (or whatever constant you chose) was sometimes recommended over COUNT(*) because poor query optimisation code would make the database retrieve all of the field data prior to running the count. COUNT(1) was therefore faster, but it shouldn't matter now.

and more interesting is a comment in that, that postgreSQL is actually faster using `COUNT(*)`!! ¯\_(ツ)_/¯

Anyway convention seems to be `COUNT(0)` in this repo so let's use that. Though do see some instances of `COUNT(*)`.

Not a biggie either way but better to follow the same convention I think.

Reply:
Done. I agree, COUNT(1) does seem to be more logical looking.